x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/user/InternalUsers.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.core.security.user;
+
+import org.elasticsearch.core.Nullable;
+import org.elasticsearch.xpack.core.security.authz.RoleDescriptor;
+
+import java.util.List;
+import java.util.Map;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+public class InternalUsers {
+
+    private record UserInstance(String name, User user, @Nullable RoleDescriptor role, Predicate<User> isUserPredicate) {
+        private UserInstance {
+            // Check that the parameters align
+            assert User.isInternal(user) : "User " + user + " is not internal";
+            // The user principal should match the official name
+            assert name.equals(user.principal())
+                : "User " + user + " has a principal [" + user.principal() + "] that dopes not match the provided name [" + name + "]";
+            // The role descriptor should match the user name as well
+            assert role == null || name.equals(role.getName())
+                : "Internal user " + user + " should have a role named [" + name + "] but was [" + role.getName() + "]";
+            // : The user object should match as an instance of the user.
+            assert isUserPredicate.test(user) : "User " + user + " does not match provided predicate";
+        }
+
+        public boolean is(User u) {
+            return isUserPredicate.test(u);
+        }
+    }
+
+    private static final List<UserInstance> INTERNAL_USERS = List.of(
```

Comment:
Good question. The answer is iterative development - I didn't know what I needed until I built it, and a list was the easier starting point.
But you're right, a Map would be better.

Reply:
Actually, there is sort of a reason...

The old code relies on `FooUser.is()`, in the interests of pure refactoring, I kept that.
Which means `findInternalUser(User)` can't use a Map lookup without assuming that `is` only returns true for users with the same principal. 
It's all part of the mess that needs refactoring.

Reply:
What I had in mind is that you'd check the result from`Map.get()` with `UserInstance.is(user)` before returning.

Reply:
I think you're right. It's Ok for `InternalUsers` to assume that the parameter `user` must have the same principal as the instance user. I was trying to avoid even building that assumption in, but there's no reason for that.